[RFC]  Floating windows in TUI and Remote UI

bfredl opened the pull request:
From the drawer of very naive and crazy stuff (one does not simply mess with window positioning and expect stuff to work). Partially based on @dzhou121 external ui stuff (crazy/naiveness mine)

do this in a script:
```
fnew
call nvim_win_float_set_pos(0,5,10,20,5)
hi Floating guibg=#000044
set winhl=Normal:Floating
set nonumber
```

bfredl committed:
floats: implement floating windows

Co-Author: Dongdong Zhou <dzhou121@gmail.com>

oni-link commented on src/nvim/ex_cmds.lua:
```diff
@@ -3003,6 +3003,12 @@ return {
+    flags=bit.bor(BANG, FILE1, RANGE, NOTADR, EDITCMD, ARGOPT, TRLBAR),
+    addr_type=ADDR_LINES,
+    func='ex_newfwin',
+  },
```
Ordering of the commands is important, see https://github.com/neovim/neovim/blob/master/src/nvim/ex_cmds_defs.h#L15-L39. Otherwise the generated array `cmdidxs` is not correct.

bfredl replied:
Also happens in #5686, ping @dzhou121 .

johnzeng commented:
hi, will this or be merged? or, still wip?

bfredl commented:
Is still WIP, many window commands can easily segfault, and the mouse is broken. Also we need a better redrawing strategy than "redraw the float everytime anything else is redrawn". Probabaly should use some buffering logic from #5686

johnzeng commented:
Good to know that. I am waiting for it.

bfredl commented:
Update: replaced fnew with api command:

```
Window nvim_open_floating_window(Buffer buffer,
                                 Integer x, Integer y, Integer w, Integer h,
                                 Integer mode, Boolean enter, Error *err)
```

mode supports anchor modes (will be relevant for UIs with different grid sizes in floats) as well making the window "unfocusable". It is still possible to make it current with api or number, but the user selecting it "by accident" with mouse or `ctrl-Wctrl-W` should be avoided.

Also added very basic mouse support.

justinmk commented:
I'd suggest "new" as the standard verb for creating things. Also in API functions we always spell "window" as `win`.

bfredl commented:
`new` isn't verb though. The alternative would be `add`, but `open` signals something more heavy (as a window or channel is) than something you just `add` to a set like a line or a highlight. `create` might also work and is slightly more general (like `nvim_create_namespace` )

justinmk commented:
`new` acts like a verb in many programming languages. It has wide recognition, it's short, strong, and familiar.

bfredl commented:
My point was primarily about operation strength, not grammar. Opening/Creating a _new_ window or buffer (or channel) is a lot heavier than adding a _new_ line (we call it [o]pen only because [n] was taken for [n]ext) or highlight, or allocating a _new_ object in "many programming languages".  These are all different verbs for a reason. (and if we're discussing common languages, `new string()` vs `string()` in javascript is just madness, while in modern c++ well-typed pointers is preferred to `new`/`delete`)

bfredl commented:
Added `nvim_add_buf()`. Note it only _adds_ a buffer to the buffer list, they are not opened/allocated until they are displayed in some window.

```
let b = nvim_add_buf("")
let w =  nvim_open_floating_window(b,5,10,20,5,0,0)
hi Floating guibg=#000044
call setwinvar(w, '&winhl', 'Normal:Floating')
call setwinvar(w, '&number', 0)
```

wsdjeg commented:
This PR is awesome, and I have a question about the floating window, I just see the doc what dese `Integer x, Integer y, Integer w, Integer h,`  base on?  computer's screen or vim's screen?

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -623,6 +623,46 @@ void nvim_set_current_win(Window window, Error *err)
+
+Window nvim_open_floating_window(Buffer buffer,
+                                 Integer x, Integer y, Integer w, Integer h,
+                                 Integer mode, Boolean enter, Error *err)
```
`nvim_open_floating_win`

justinmk commented:
> nvim_add_buf

Here "new" as the verb makes sense to me: `nvim_new_buf()`. We're creating an object. The buffer list is essentially a vtable, an implementation detail, so "add" isn't the characteristic operation.

bfredl commented:
The buffer list most definitely not an implementation detail: it is well known and used vim functionality that a buffer can be "listed" without actually being "opened". The API functions _current_ operation is only to add/return a number in the buffer list, but "add" might not be the best either, as it can return an existing buffer. 

However, that is not _necessarily_ the best operation, we could have a function that only creates a new (unnamed) buffer and fully allocates it (this special case might not need to make it temporarily displayed, as there is no swapfile, encoding errors, and probably we want no autocmds either) The current behavior makes sense in the "floating window" case given that it will immediately _opened_ in a window anyway, at which point all the vim stuff happens.

ZyX-I commented on src/nvim/buffer_defs.h:
```diff
@@ -1154,6 +1172,9 @@ struct window_S {
   int w_tagstacklen;                    /* number of tags on stack */
 
+  int w_floating;                       /* if the window is floating */
+  FloatMode w_float_mode;
```
You can’t hold `kFloatAnchorNW | kFloatStandalone` in `FloatMode`, need generic integer type.

ZyX-I commented on src/nvim/buffer_defs.h:
```diff
@@ -1154,6 +1172,9 @@ struct window_S {
   int w_tagstackidx;                    /* idx just below active entry */
   int w_tagstacklen;                    /* number of tags on stack */
 
+  int w_floating;                       /* if the window is floating */
```
Should not that be boolean? Also comment is wrong, need `///<`.

ZyX-I commented on src/nvim/screen.c:
```diff
@@ -419,18 +419,29 @@ void update_screen(int type)
-        did_one = TRUE;
-        start_search_hl();
+
+  for (int floating = 0; floating < 2; floating++) {
```
AFAIK “two-iteration loop” is supposed to be an antipattern.

ZyX-I commented:
I have a suggestion: either make `x` and `y` `Float` and allow negative and fractional values, or add another argument which specifies what `x` and `y` are related to and what they actually mean. Neither makes sense for TUI, but GUI may open windows outside of the window.

*Or* do the other thing: `open_floating_window` already contains lots of arguments. Make that dict:

```C
/// Open floating window
///
/// @param[in]  buffer  Buffer to display in that window.
/// @param[in]  position  Window position, pair of two integers which may be 
/// negative.
/// @param[in]  options  Additional options:
///                      Option | Type | Description
///                      ------ | ---- | -----------
///                      anchor | String | Make position be relative to the 
///                             |        | given anchor: "type-pos" where "type"
///                             |        | is one of "screen", "editor_area",
///                             |        | "wm_window" and "pos" is one of "NW",
///                             |        | "NE", "SW" or "SE"; additionally
///                             |        | "cursor" position is supported.
///                             |        |
///                             |        | Defaults to "cursor".
///                             |        |
///                             |        | TUI only supports "editor_area-…" and
///                             |        | "cursor".
///                      dimensions | pair of Integers | Dimensions in screen
///                                 |                  | cells, width x height.
///                                 |                  |
///                                 |                  | Defaults to either just
///                                 |                  | enough to show the
///                                 |                  | whole buffer contents
///                                 |                  | (not updated when
///                                 |                  | buffer contents
///                                 |                  | changes, call
///                                 |                  | `nvim_reconfigure_floating_window({dimensions: NIL})`
///                                 |                  | then), or just enough
///                                 |                  | to occupy all available
///                                 |                  | space, whichever is
///                                 |                  | lesser.
///                      standalone | Boolean | Make window standalone.
///                                 |         |
///                                 |         | Defaults to false, may be
///                                 |         | ignored in TUI.
///                      focusable | Boolean | Make window not focusable, and
///                                |         | also ignore mouse events.
///                                |         |
///                                |         | Defaults to false.
///                      position_type | String | Position type: either "cells"
///                                    |        | (position in display cells) or
///                                    |        | "pixels".
///                                    |        |
///                                    |        | Defaults to "cells" for
///                                    |        | "editor_area-*" and "cursor"
///                                    |        | anchors, "pixels" otherwise.
Window nvim_open_floating_window(Buffer buffer, ArrayOf(Integer) position,
                                 Dictionary options, Error *const err)

/// Reconfigure floating window
///
/// @param[in]  window  Window to reconfigure.
/// @param[in]  options  @see nvim_open_floating_window()
///                      Missing values mean “do not change option”, use NIL to 
///                      force defaults.
void nvim_reconfigure_floating_window(Window window, Dictionary options,
                                      Error *const err)
```

justinmk commented:
> The buffer list most definitely not an implementation detail: it is well known and used vim functionality 

It's an irrelevant detail in the context of the API. When one creates a buffer one does not think "I want to add a buffer to the buffer list", only "I want a new buffer".

bfredl commented:
Again, the name describes what the function does _right now_, which was the simplest thing to implement. The relevant change would be to change what the function actually does (fully allocate a buffer), which will be more involved but probably worth it in the long run.

bfredl commented:
Update: does not crash on `:q<cr>`, neither in floating window nor last non-floating window. However the last is guaranteed _only if_ there is no other tabpage. Interaction with multiple tabpages in general is NOT going to be pretty. Especially with external floats which one would expect not to disappear when switching tabpage...

Also, `nvim_create_buf` works, but modification must be made to `nvim_buf_set_option` to accept it before buffer is visited.

Current state of art:
```
let b = nvim_create_buf(v:false)
call nvim_buf_set_option(b, "buftype", "nofile")
call nvim_buf_set_lines(b,0,-1,v:true,["text"])
let w =  nvim_open_floating_win(b,5,10,20,5,0,0)
hi Floating guibg=#000044
call setwinvar(w, '&winhl', 'Normal:Floating')
call setwinvar(w, '&number', 0)
```

bfredl commented:
Update: an external ui can now opt-in to receive float draw events relative a separate grid. However they currently receive no context what the grid means/ how it should be placed, I will fix that soon. Can be tested with
https://github.com/neovim/python-gui/compare/master...bfredl:multigrid?expand=1 which currently just allocates a new top-level window with no cue of positioning. But I think floats should only be used for ad-hoc single purpose windows/buffers, rather I have a separate branch for multiple long-lived "normal" top-level windows I will publish "soon".

bfredl commented:
Copied from #7541 :

The multi-grid mode can be tested with the multigrid branch of python-gui https://github.com/neovim/python-gui/pull/36. The python-gui code is a big mess though because I changed my mind 5 times how to communicate grid/cursor focus to the UI. I might clean it up later. The rule I ended up with was to manage grids together with cursor position: the most recent `set_grid` tells the cilent the current grid, and together with the most recent `set_cursor` forms the 3-tuple `(grid_id, row, col)`which tells where drawing is done, and this state at the end of a `"redraw"` array of updates should be seen as the current focus. Perhaps the event should be changed to a joint `grid_cursor_goto(grid_id, row, col)` to make this explicit. Also `resize` and `scroll` operates on the currently focused grid.

bfredl commented:
added `float_info` and `float_close` events

bfredl commented:
Apart from going through all wincmd:s and make them not crash, there are some open questions:

- Should  wincmds (like `<c-w>J` be able to implicitly turn a float into a non-floating split window) ?
- Should there be a wincmd to turn a window into a float (most useful for a standalone float in external ui)?
- Should there be an option to implicitly make "special" windows like quickfix, preview and cmdline window into floats (either all at once, or a flag for each). Or should this be handled by autocmds?

bfredl commented:
Wincmds that have been reviewed (will be updated in place):

- [ ] `CTRL-W s :split`
- [ ] `CTRL-W v :vsplit`
- [ ] `CTRL-W n :new`
- [ ] `:vnew`
- [ ] `CTRL-W ^`
- [x] `CTRL-W q :quit`
- [ ] `CTRL-W c :close`
- [ ] `:hide`
- [ ] `CTRL-W o :only`
- [ ] `CTRL-W j`
- [ ] `CTRL-W k`
- [ ] `CTRL-W l`
- [x] `CTRL-W w`
- [x] `CTRL-W W`
- [ ] `CTRL-W t`
- [ ] `CTRL-W b`
- [ ] `CTRL-W p`
- [ ] `CTRL-W P`
- [ ] `CTRL-W r`
- [ ] `CTRL-W R`
- [ ] `CTRL-W x`
- [ ] `CTRL-W K`
- [ ] `CTRL-W J`
- [ ] `CTRL-W H`
- [ ] `CTRL-W L`
- [ ] `CTRL-W T`
- [ ] `CTRL-W =`
- [ ] `CTRL-W - :resize -N`
- [ ] `CTRL-W + :resize +N`
- [ ] `CTRL-W _ :resize N`
- [ ] `CTRL-W < :vertical resize -N`
- [ ] `CTRL-W > :vertical resize +N`
- [ ] `CTRL-W | :vertical resize N`

bfredl commented:
Update: most wincmds do not longer crash at least in "basic" operation

I make the distinction between two types of floats: "basic" floats that always should be drawn on top of the global grid (ext ui can still draw them with different font size etc) , these cannot be moved/attached with user wincmds (but they can be closed), And "standalone" floats that would be reasonable to create directly by user action (for "special" windows like help, preview etc), and thus make sense to manipulate with user commands, and can be "attached" using `CTRL-W J` etc and `CTRL-W T`.  These make the most sense in external UI, which can draw them as top-level windows that are separately movable and resizable. But for testing purposes they also work in TUI, and are then drawn the same way as basic floats, but can be moved/resized with wincmds.

Tehnix commented:
Just to see if I understood this correctly, would this allow a plugin to create e.g. tooltips by using floating windows?

Also, is there any thing left that needs to be implemented? It seems all things mentioned have been taken care of, as per the fully checked off wincmd list and the last comment.

bfredl commented:
Tests are needed, as well as proper documentation, including of the "multigrid" protocol extension.

There are lots of code paths in vim that end up in window layout being changed. But probably some general testing on master branch will be needed to catch them all...

bfredl commented:
Update: multigrid supports basic mouse events, using the extended format `<LeftMouse><gridnr,col,row>`.  You can click _inside_ a standalone float and have it focused. However, yet there is currently no way a gui can say "float was focused" (by clicking windowbar or alt-tab)  _without_ a specific position, probably a separate `<GridFocus><gridnr>` event is needed; it can still follow mouse codepaths.

wsdjeg commented:
@bfredl I just test this branch loadl, and there is no `fnew` command.

bfredl commented:
I'm not sure how to best balance flexibility/precisiton of postition specification and simple sensible defaults. For standalone floats I think it is simple, make unspecified (default) position mean that we leave positioning up to the GUI, which can just rely on the WM placing the window in a sensible position.

But for anchored (non-standalone) floats I guess that some plugins do not want to specify the position with exact coordinates, rather want to to say "place float in some sensible way close to screen cell (row,col), preferably not covering the row (typically the cursor row) at all", with similar heuristics to how popupmenu is placed. Or maybe even things like "place it centred on top of screen or specific window". But I'm not sure if these calculations should be done in nvim core or rather should be forwarded to external GUI. The later would mean things like adding a "auto" or "none" anchor mode to the options dict. It also depends on to what extent we  consider the positions dict an almost pixel-exact specification (when present) or more of a "general recommendation" the GUI can ignore/reinterpret when it feels it has a better idea.

"Ideal" layouting might necessitate a more involved procedure negotiating not only position but also size between the gui and nvim core/the plugin, but is probably out of scope for this PR.

Input from GUI (and plugin) developers would be very welcome here. Also I'm not very good at writing detailed technical documentation, but I could help with making a "reference implementation" in one of the GUI projects that are interested in this idea. Ping @dzhou121 @equalsraf @rhysd @bryphe @chilee @lunixbochs (and everyone else implementing GUIs and embeddings)

badeip commented:
I haven't followed neovim development for some time, so the following might be irrelevant and/or out of scope for this PR.
Regarding layout, couldn't UI elements be described by placing them in a 10000x10000 matrix.
Lets say that the actual dimensions of your terminal is, at the moment, 60x120:
This means that one TUI line is 166 units (10000/60).

The "edit widget", _can_ be located at x = 0, y = 0, and have a w = 10000, h = 9668.
The "status line widget" _can_ be located at x = 0, y = 9668, and have w = 10000, h = 166
The "command line widget" _can_ be located at x = 0, y = 9834, and have w = 10000, h = 166.
To allow stacking (floating windows), each widget should have a layer indicator.

This scheme can be used for GUIs where the location within the matrix is converted to actual pixel values, so that we have an uniform method to describe both the TUI and GUI.

Something like this can be saved in shada(?), and programatically fetched- and updated (the CML/statusline can e.g be moved to the top of the window) with appropriate functions.

ZyX-I commented:
If it is configurable, it must not be ShaDa, that is not for user configuration. If it is not, it is a bad idea. I personally would not bother with anything, but simple monospace font grid and exact positions and dimensions, but the result is les then perfect for GUI. And definitely not invent some kind of universal scalable grid like you suggest as it is simply impossible: that works as long as size differences do not exceed ~10-20%, but shrink or expand enough and user will start wanting to either shift and expand floating window as shrinked variant became too narrow, or prevent the window expansion above certain level as wide texts are harder to read.

Also “status line widget” having height of 166 is nonsense, with the current implementation you need status line to have *exactly* one line, *regardless* of what size rest of the UI is. Though even if we supported multiline statuslines I would still want the same thing: first, it would be quite some time before plugins adapt, and not all of them will bother to as Bram probably won’t merge the patch (there *is* multiline statusline patch created by vim-jp user group, though I am not sure whether they tried to make it into upstream). Second, I simply don’t know what I would want to see on the second line.

And, BTW, your suggestion sounds like an optimisation for the computer without floating-point numbers support. You *don’t* need any grid like that to express “I want UI element at top-center”, just make the position have type `double[2]` with each value in range `[0, 1]` (inclusive).

bfredl commented:
For TUI we should simply disallow everything but (integer) cell addressing. Also for GUI I suppose floating point cell positioning would be good as a MVP. We can restrict positioning configuration in the first merge of this feature, then after GUI people has worked with this feature we can use their feedback to correctly design more flexible addressing modes. Unfortunately I have been to busy with work and other stuff the last weeks, but I hope to finish the MVP sometime soon...

[re]configuring the layout of existing elements are out of scope of this feature. Floating windows are  for plugins implementing new screen elements. _Possibly_ I could imagine reimplementing more rich TUI completion popupmnu (and/or vim 8's termballoneval, which reuses pum code paths) as floating window in vimL/lua, but that should be a follow-up PR.

wsdjeg commented:
> Possibly I could imagine reimplementing more rich TUI completion popupmnu (and/or vim 8's termballoneval, which reuses pum code paths) as floating window in vimL/lua, but that should be a follow-up PR.

greate! I also  want this feature.

bfredl commented:
Update: For inline floats I restricted position to only cell units. Still with anchoring in four different corners, and with both `editor` and `cursor` relative, but `cursor` coordinates are translated to `editor` screen coordinates to the GUI. Standalone floats cannot have position specified at all.

I implemented this limited positioning in python-gui at neovim/python-gui#37.
Simple demonstration of both inline and standalone floats
![screenshot](https://user-images.githubusercontent.com/1363104/34652812-832b7b38-f3e3-11e7-8fce-734cfad39de6.png)

I don't think we want to formalize styling information in this PR. But we could add some kind of `user` dict to options, forwarded from plugin to gui, and develop (on wiki) a list of recommended keys for thinks like

- (optional) border thickness and color
- relative font size hint
- or even separate font family
- opacity of background

bfredl commented:
Update: rather than fixing all weird redraw issues in single-grid mode (stopping statuslines from overdrawing floats seems to be positively impossible) I plan to immediately do as I suggested earlier on gitter: `screen.c`,`window.c` will _only_ support multi-grid mode. Instead, `ui_compositor.c` will be implemented in analogy with `ui_bridge.c`, which will translate from the multigrid protocol back to the current (single-grid) protocol, for the UIs that don't opt-in to the new protocol. Even a very crude and simple "compositing" algorithm will probably end more efficient, as we easily can avoid _a lot_ of window/buffer redraws that single-grid mode would need to do conservatively.

The branch is currently [here](https://github.com/bfredl/neovim/tree/floating_compositor), it will replace this one as soon as it is at least as functional...

bfredl commented:
Update: redrawing in TUI should now be more correct and effective (with the glaring exception of scrolling, I will fix that soon). Also, moving and closing floating windows can be done without redrawing the underlying window (just retransmitting the screen grid). Using `'writedelay'` should illustrate the difference, as it only slows down redrawing, not retransmission.

Also, can now support TUI drawing of floats and externalized floats in GUI at the same time. Not that it is very useful by itself, but it is rather a long-term goal that it should be possible to attach a TUI to nvim instance with "rich" UI, and get a reasonable TUI view without collapsing the GUI down to only TUI features. Later on, also the popupmenu and cmdline could be rendered on separate layers, and thus `ext_popupmnu` and `ext_cmdline` could degrade the same way (as was discussed back in  #4432 ).

brcolow commented on runtime/doc/ui.txt:
```diff
@@ -287,7 +302,7 @@ Only sent if `ext_cmdline` option is set in |ui-options|
 	The Nvim command line can be invoked recursively, for instance by
 	typing `<c-r>=` at the command line prompt. The `level` field is used
-	to distinguish different command lines active at the same time. The
+	to distinguish different command :slines active at the same time. The
```
Is this a typo?

brcolow commented on src/nvim/api/vim.c:
```diff
@@ -690,6 +690,97 @@ void nvim_set_current_win(Window window, Error *err)
+/// Create new buffer with requested buftype
+///
+/// 
+/// For instance, swap file prompt is only shown then.
```
Copy paste error on this line?

bfredl replied:
Yes, line was only relevant when function could open existing files.

bfredl commented:
Experimented with semi-transparent floats in TUI, now that we can do compositing:

![screenshot from 2018-02-25 16-36-04](https://user-images.githubusercontent.com/1363104/36643354-497e731c-1a4a-11e8-9cfc-3103b634d877.png)

Not sure how much sense it really makes, but more efficient and prettier (blending instead of just override color) `cursorline`/`column` could be implemented with similar code in compositor module.

But back to wincmd tests...

WouterSioen commented:
Any update on this?

bfredl commented:
As the commit log tells I'm working on wincmd tests (and in the process fixing bugs), slowly but regularly...

bfredl commented:
This branch now depends on #8221 and #7992. Which makes "TUI mode" much simpler: multiple grids can b composed using `memcpy()` as it should be with no state other than the grids itself (including their position on single TUI grid). Though state could be added for efficiency reasons, for instance with emulated scroll we could turn off direct drawing and only transmit the final state in `ui_flush()`

ELLIOTTCABLE commented:
I'm confused — what's the blocking pattern, here?

It looks like implementation of *Terminal* floats (which sounds like a Hard Problem) is blocking support for *GUIs* like VimR from implementing (much easier) GUI floats / hovertext / etc etc etc.

Is there a chance of splitting the API-work here, out from the TUI implementation of said API? Maybe publish something that the VimR (qvacua/vimr#535), et al (daa84/neovim-gtk#113) can implement in parallel?

It's quite possible I'm missing something here. I hope this is a welcome input! 🤣

bfredl commented:
The honest answer is my lazyness and inability to keep focus, though I'm currently working on finishing #8221 which this depends on.

> It looks like implementation of Terminal floats (which sounds like a Hard Problem) is blocking support for GUIs like VimR from implementing (much easier) GUI floats / hovertext / etc etc etc.

This part I fixed in march/april, the TUI is already more or less an "external UI" with respect to floats (almost, the handling is still done in the main thread than the TUI thread because it was easier that way, but TUI stuff has been separated from screen.c which is what is important)

The most non-trivial part, as already mentioned, is instead the handling of wincmds, because floats breaks all assumptions of the existing frame layout.

ELLIOTTCABLE commented:
Hm. Maybe I'm missing something fundamental — I re-read the thread above, but I still don't understand how this is fundamentally blocked? I mean, we had balloonexpr, and it was already working, prior to the fork. All the new fundamental work is definitely valuable in the long-run, but I'd *really* like to see support for mouse-hover balloons restored, at least for NeoVim-backed GUIs, ASAP.

idk! You're the one doing all the work, and it's in your free-time, so I can't really complain; just thought you might want to know what there's demand for …

bfredl commented:
I don't think I have said porting `ballonexpr`/`termballoneval` is fundamentally blocked, on gitter I have said quite the contrary (we can use vim implementation for now, and rebase on this later), nevermind my rant about useless positioning limitations, someone willing to spend time on porting the patches could also look into fixing that (or not, this is up to this hypothetical patchporter, which won't be me)

janlazo commented:
I can port the patches but I'd like some hints on which code to restore.

bfredl commented:
@janlazo I guess you could follow the features `FEAT_BEVAL`, `FEAT_BEVAL_GUI`, the main code being in `src/beval.c`. Then `FEAT_BEVAL_TERM` would be the code introduced by patches some months ago for terminal support.

Piping commented:
@bfredl I tried your code! It's awesome!
Is there a way to adding shadow/different color border for the floating windows?

![image](https://user-images.githubusercontent.com/12042284/42131040-a9c81c7c-7cac-11e8-9bd1-9367e1487665.png)

bfredl commented:
> A way to attach floating windows back to ground windows, E.g. Ctrl-W H

This is already possible, use `ctrl-W J` etc. In general, this first PR will be quite low-lovel, I hope to add both more useful builtins and runtime functionaly for plugins/users later. For instance, similar to the shared "preview window", there should probably be a shared abstraction for the tooltip/signature window, so a plugin like vim-jedi can show a buffer with the signature with a single or two function calls.

chemzqm commented:
This feature would also really useful for workspace symbol search, since the search happens in server, and I can't use existing plugins to do that.
https://code.visualstudio.com/assets/docs/getstarted/tips-and-tricks/go_to_symbol_in_workspace.png

Piping commented:
One crashed case found. https://github.com/junegunn/fzf.vim/issues/664

bfredl commented:
@piping thanks, I will rebase and fix it as soon as I'm done with #8221 (quite soon)

Shougo commented:
I may use the feature in denite.nvim.

vrinek commented on runtime/doc/ui.txt:
```diff
@@ -170,6 +171,14 @@ option by itself will never cause any additional grids to be created.
+TODO: rewrite:
+By default, Nvim will only draw windows on one single grid, just like a
+terminal emulator. But when multigrid mode is activated Nvim might use
+multiple grids. Currently this is only used for floting windows |ui-float|,
```
`floting` typo?

bfredl replied:
Thanks. Most of this paragraph is obsolete...

marxin commented:
Thank you very much for working on this. Looks the WIP pull request has been opened for quite some time. Can please anybody explain what's remaining to have in neovim master? Thanks!

justinmk commented:
@marxin The discussion here is already too long, please stay on topic. People are pushing regular updates to this and related PRs, that should be enough of a "status update".

bfredl commented:
@marxin I just rebased this on top of #8455. There is some remaining rough edges to sort out, like making the external UI API for both these features more consistent, as well as both working correctly with tabs and multiline messages. I think it would be reasonable to get #8455 and this PR merged soon after 0.3.2 is released, with the understanding that these features remain experimental for a while during the 0.3.3 cycle.

jamessan commented:
> neovim version 0.3.3 is already released.

That comment was made before the release numbering was changed.  Releases from master will follow the x.y.0 numbering, so we can do bug fix releases (like 0.3.3 was).

That should say "during the 0.4.0 cycle" now.

Shougo commented:
Oh.  OK.

bfredl commented:
@Shougo I am working on merging it, but in a piece-by-piece fashion.

- #9434 extracted some cleanups and fixes
- #9429 contains most of the mouse code, which is useful also for multigrid.

bfredl commented:
@breja We need to update the API so that it works with `ext_multigrid` out of the box, i e support window-relative positioning, not just relative to the global grid. Probably we should just turn all positioning into the options dict (this will sanely allow different positioning modes later, such as relative to a text position in a buffer-window). If we do multigrid-awareness correctly, it should work with `ext_windows` also, when that becomes available.

justinmk commented on runtime/doc/ui.txt:
```diff
@@ -636,5 +637,29 @@ Only sent if `ext_wildmenu` option is set in |ui-options|
+
+This extension allows the UI to draw floating windows (described at
+|nvim_open_float_win|) by receieving updates on separate grids. Using this
+extension implicitly activates `ext_newgrid` extension. Floats will be drawn
```
IMO it would be preferable to "require" `ext_newgrid` (error if missing), rather than implicitly enabling stuff. Otherwise the list of enabled extensions does not match the explicit list sent by the client.

bfredl replied:
lets integrate the docs with the "multigrid" selection. The floats feature can be relegated to a debug flag only, as we discussed on gitter.

justinmk commented on src/nvim/buffer_defs.h:
```diff
@@ -958,6 +958,33 @@ struct matchitem {
+
+    kFloatAnchorNW = 0,
+    kFloatAnchorNE = 1,
+    kFloatAnchorSW = 2,
```
`kFloatAnchorEast`/`kFloatAnchorSouth` also use 1/2. Can't `win_config_float()` use `kFloatAnchorNE`/`kFloatAnchorSW` instead?

bfredl replied:
It doesn't mean the same thing, one are flags, the other are combinations of flags. Though we could use only concrete flags in code like `kFloatAnchorEast | kFloatAnchorSouth` and only keep dummy combination item for compatibility. Or not use `enum` storage for values, rather typedef an int.

justinmk commented on src/nvim/window.c:
```diff
@@ -6086,20 +6489,26 @@ void win_findbuf(typval_T *argvars, list_T *list)
+win_T *lastwin_nofloating(void) {
+  win_T *res = lastwin;
+  while (res->w_floating) {
+    res = res->w_prev;
```
Maybe it's worth  a sanity check:

    assert(res != NULL && res != firstwin);

justinmk commented on src/nvim/window.c:
```diff
@@ -505,6 +536,217 @@ static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
+  char *str = anchor.data;
+  if (!STRICMP(str, "NW")) {
+    *out = kFloatAnchorNW;
+  } else if (!STRICMP(str, "NE")) {
```
could use `striequal`

justinmk commented:
- Crash found with these steps (`:tabclose` on two overlapping floats):
  ```
  nvim -u NORC
  :call nvim_open_float_win(bufnr('%'), v:true, 60, 20, {'relative':'editor','row':8,'col':20})
  :call nvim_open_float_win(bufnr('%'), v:true, 60, 20, {'relative':'editor','row':8,'col':30})
  :tabnew
  :tabnext
  :tabclose
  ```
- Do floats deal with overlap explicitly or only accidentally? Navigating between overlapping floats does not bring the current window to the "foreground".
  ```
  :call nvim_open_float_win(bufnr('%'), v:true, 40, 8, {'relative':'editor','row':8,'col':20})
  :call nvim_open_float_win(bufnr('%'), v:true, 40, 8, {'relative':'editor','row':8,'col':30})
  ```
- Is it feasible to deal with overlap in future PR (if so I'll create a ticket)? Something like a z-index?
- Currently floats have "configuration" such as `unfocusable`, `standalone`, etc. Is there a strong reason these couldn't be a window-local option?
  - `win_config_float` is useful _internally_, of course.
  - Window could have a single dict option like this:
    ```
    :setlocal float={'unfocusable':0,'standalone':0,...}
    ```
    - This could "promote" a window to a float and vice-versa.
  - I'm willing to work on this if wanted.
- Instead of the name "standalone", what about "flex"? Other shorter/more-familiar names: "orphan", "external", or "unmanaged".

jamessan commented:
>     * Window could have a single dict option like this:
>       ```
>       :setlocal float={'unfocusable':0,'standalone':0,...}
>       ```

Wouldn't this be the first dict option?  What's wrong with using the canonical comma-separated, colon-delimited format for other complex options?

Typing a dict for an option is going to be error-prone, since people are likely to want to insert whitespace but that's not allowed (without going `\`-happy) for `:set`

justinmk commented:
> Wouldn't this be the first dict option? What's wrong with using the canonical comma-separated, colon-delimited format for other complex options?

Doesn't matter too much in this context, unless it turns out to be easier to implement. Anyways, let's save that discussion for later.

bfredl commented:
> Do floats deal with overlap explicitly or only accidentally?  Is it feasible to deal with overlap in future PR (if so I'll create a ticket)? Something like a z-index?

There is a z-index internally, but it is not updated for anything yet. Bringing focused float to front makes sense to start with, I will add it right away.

> Window could have a single dict option like this ... or comma separated list:

This means that the plugin has to use string interpolation to configure the float, rather than construct the dict using the native language construct i e `{anchor='NW', relative='buffer', lnum=pos[1], col=pos[2], unfocusable=true}` in lua. Also the semantics would be weird with `setl`; we could disable `setl` for the option i e window-only local option, but that would be yet a new option scope kind[1]. Possibly there could be single option for quick checking the mode like ` float` / ` nofloat`, or maybe even `float=standalone` (or whatever we call it), but the concrete configured position is not an option, just as the current split layout is not an option (though there is some options that constrain it).

[1] Unless it already exists some option like that, wouldn't surprise me anymore.

justinmk commented:
> but the concrete configured position is not an option, just as the current split layout is not an option

That's true, window layout is not a "window option". Though it could be viewed as a "tab option" (where "option" is understood as Vim-parlance for "property").

It may be awkward to force it into the "options" space, I guess.

At the very least I hope we can avoid treating "floating windows", "flex windows", etc., as different classes of things, and instead stick with just _windows_ which have _layout properties_. So `nvim_win_config` would operate on any window, we wouldn't have a separate `nvim_win_config_float`, `nvim_win_config_xx`, etc. 

Does this seem feasible?

bfredl commented:
Yes, general `nvim_open_win` and `nvim_win_config` functions as we discussed make sense, but only supporting floats initially in this PR.

chemzqm commented on src/nvim/api/window.c:
```diff
@@ -432,3 +432,41 @@ Boolean nvim_win_is_valid(Window window)
+/// See documentation at |nvim_open_float_win|, for the meaning of parameters.
+///
+/// When reconfiuring an existing float, absent option keys will not be
+/// changed. The following restriction apply though: `x`, `y` and `relative`
```
There is `row` and `col` instead of `x`, `y` at documentation of `nvim_open_float_win`, so what should be used?

bfredl replied:
row, col. I'm in the process of updating it

chemzqm commented:
Can't move floating window without change content of buffer:
``` vim
let s:buf = nvim_create_buf(1)

let s:win = nvim_open_float_win(s:buf, v:false, 5, 5, {
      \ 'col': 10,
      \ 'row': 1,
      \ 'unfocusable': v:true,
      \ 'standalone': v:true,
      \})
call nvim_buf_set_lines(s:buf, 0, -1, v:false, ['a', 'b', 'c', 'd', 'e'])
call setwinvar(s:win, '&signcolumn', 'no')
call setwinvar(s:win, '&number', 0)
call setwinvar(s:win, '&relativenumber', 0)
call setwinvar(s:win, '&winhl', 'NormalNC:Pmenu')

function! s:config()
  "call nvim_buf_set_lines(s:buf, 0, -1, v:false, ['a', 'b', 'c', '', ''])
  call nvim_win_config_float(s:win, 5, 5, {
        \ 'col': 11,
        \ 'row': 1,
        \})
endfunction

call timer_start(1000, {-> s:config()})
```
source these script with `nvim -u NORC`,  the first column would disappear, it works as expected when content of buffer changed.

chemzqm commented:
It's caused by `CursorHoldI` triggered by coc, it should be removed.

bfredl commented:
@chemzqm Thanks for testing! Redrawing is still a bit unreliable, I probably messed up something in latest rebase. Hopefully I will have some time in the weekend to patch things up.

bfredl commented:
Updated the API (and docs) to use non-float specific names. Added 'window' relative positioning, which is needed for multigrid support. The end goal is to support lnum+col positioning, though that looks a bit messy, so it will be a follow up PR. For now `cursor` relative positioning works (i e one can use row=1 and col=0 to show float right below cursor), and multigrid client will receive grid-relative position when using `relative=cursor`.

Tests now use proper multigrid representation. Still a few failures due to leaks and missing checks (external window should only be valid with multigrid).

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -987,6 +987,78 @@ Buffer nvim_create_buf(Boolean listed, Boolean scratch, Error *err)
+/// Exactly one of `external` and `relative` must be specified.
+///
+/// @param buffer handle of buffer to be displayed in the window
+/// @param enter whether the window should be entered (made the current window)
```
Vim parlance might be `current` (thinking of `NormalNC` etc.). Or else `focus` (in analogy with `FocusGained/Lost`), which is also more common.

bfredl replied:
focus as opposed to _what_? "current" is our jargon for current window (`nvim_get/set_current_xx`). "enter" is our jargon for entering a window/buffer (`BufEnter` / `WinEnter`).

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -987,6 +987,78 @@ Buffer nvim_create_buf(Boolean listed, Boolean scratch, Error *err)
+///        "SW" south-west
+///        "SE" south-east
+///     `unfocusable`: The window will not be focused by wincmds and
+///         mouse events. It can still be made current by API calls.
```
Negative logic is confusing, so `focusable` is a better name. It's ok if the default is `true`.

bfredl replied:
"made current" means "made the current window", "entered" would be established verb for this. "focused" means entered due to user input being directed towards the window, i e by clicking inside the window.

justinmk commented on src/nvim/api/window.c:
```diff
@@ -432,3 +432,41 @@ Boolean nvim_win_is_valid(Window window)
+/// floating and external windows (including changing a split window to these
+/// types).
+///
+/// See documentation at |nvim_open_window|, for the meaning of parameters.
```
tag `|nvim_open_window|` won't exist, should be `|nvim_open_window()|`

ilAYAli replied:
```suggestion
/// See documentation at |nvim_open_win()|, for the meaning of parameters.
```

s/window/win/

justinmk commented on src/nvim/screen.c:
```diff
@@ -449,7 +457,16 @@ void update_screen(int type)
+
   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if (wp->w_redr_type == CLEAR && wp->w_floating && wp->w_grid.chars) {
+      // TODO(bfredl): this is overly expensive. The solution is to CLEAR less
```
could add a field to `g_stats` so we have a counter of CLEARs which could be (eventually) tested.

bfredl replied:
It could be this TODO was from before the clearmaggedon. Now basically only resize and user intervention should lead to a clear.

justinmk commented on test/functional/ui/screen.lua:
```diff
@@ -1146,7 +1160,9 @@ function Screen:redraw_debug(attrs, ignore, timeout)
     assert(method == 'redraw')
     for _, update in ipairs(args) do
-      print(require('inspect')(update))
+      if update[1] ~= "mode_info_set" then
```
is this because mode_info_set is noisy? Would be good to add a comment.

bfredl replied:
indeed.

Shougo commented:
Floating window support is here in denite.nvim.

https://github.com/Shougo/denite.nvim/pull/580

wsdjeg commented:
I will also add floating windows support in flygrep :), but the api seems has been changed.

https://github.com/SpaceVim/SpaceVim/pull/2216

chemzqm commented:
Could floating window support optional border? 
I can draw the border as buffer content, but it doesn't looks well when the buffer contains more lines than window height?

bfredl commented:
Yes, but that would be a follow-up PR. `winhl=Normal:GroupWithBG` works for now to make the float stand out.

bfredl commented:
Implemented basic z-order: if float becomes current window, it will be put on top, with the exception of popupmenu which always is on top if shown. It can be made more fancy later on.

chemzqm commented:
I think a function for close preview window or close window without focus is needed.

I can close the preview window by `wincmd c` but it requries focus the preview window, the problem is  when the user have selected region at this time the change of focused window break current selection.

bfredl commented:
`:[winnr]close` can be used, but it is not very convenient. We could add `nvim_win_close([winid])` which works the same but with winid.

bfredl commented:
Working on the last issues, ci build is aalmost passing now. I plan to merge this today or tomorrow. It is usable enough as a MVP (as demonstrated by plugins adding support already). The API/functionality (and tests with them) can be expanded in follow-up PRs.

bfredl commented:
Merged. Sorry for taking so long. I will lock this thread, as dealing with overlong github threads is a mess. Requests for additions/improvements (or bugs too minor to deserve their own issue) can be in #9421. Feel free to repeat stuff that is hidden somewhere in the middle of this thread and still are a concern.
